src/server.rs

```diff
@@ -0,0 +1,409 @@
+            }
+            slice::from_raw_parts(cert_chain, cert_chain_len as usize)
+        };
+        let private_key: &[u8] = unsafe {
+            if private_key.is_null() {
+                return NullParameter;
+            }
+            slice::from_raw_parts(private_key, private_key_len as usize)
+        };
+        let mut private_keys: Vec<PrivateKey> = match pkcs8_private_keys(&mut Cursor::new(private_key)) {
+            Ok(v) => v,
+            _ => match rsa_private_keys(&mut Cursor::new(private_key)) {
+                Ok(v) => v,
+                Err(()) => return rustls_result::PrivateKeyParseError,
+            }
+        };
+        let private_key: PrivateKey = match private_keys.pop() {
+            Some(p) => p,
+            None => return rustls_result::PrivateKeyParseError,
+        };
+        let parsed_chain: Vec<Certificate> = match certs(&mut cert_chain) {
+            Ok(v) => v,
+            Err(()) => return rustls_result::CertificateParseError,
+        };
+        match config.set_single_cert(parsed_chain, private_key) {
+            Ok(()) => rustls_result::Ok,
+            Err(e) => map_error(e),
+        }
+    }
+}
+
+/// Turn a *rustls_server_config_builder (mutable) into a *rustls_server_config
+/// (read-only).
+#[no_mangle]
+pub extern "C" fn rustls_server_config_builder_build(
+    builder: *mut rustls_server_config_builder,
+) -> *const rustls_server_config {
+    ffi_panic_boundary_ptr! {
+        let config: &mut ServerConfig = try_ref_from_ptr!(builder, &mut ServerConfig,
+             null::<rustls_server_config>());
+        let b = unsafe { Box::from_raw(config) };
+        Arc::into_raw(Arc::new(*b)) as *const _
+    }
+}
+
+/// "Free" a server_config previously returned from
+/// rustls_server_config_builder_build. Since server_config is actually an
+/// atomically reference-counted pointer, extant server_sessions may still
+/// hold an internal reference to the Rust object. However, C code must
+/// consider this pointer unusable after "free"ing it.
+/// Calling with NULL is fine. Must not be called twice with the same value.
+#[no_mangle]
+pub extern "C" fn rustls_server_config_free(config: *const rustls_server_config) {
+    ffi_panic_boundary_unit! {
+        let config: &ServerConfig = try_ref_from_ptr!(config, &mut ServerConfig, ());
+        // To free the server_config, we reconstruct the Arc. It should have a refcount of 1,
+        // representing the C code's copy. When it drops, that refcount will go down to 0
+        // and the inner ServerConfig will be dropped.
+        let arc: Arc<ServerConfig> = unsafe { Arc::from_raw(config) };
+        let strong_count = Arc::strong_count(&arc);
+        if strong_count < 1 {
+            eprintln!(
+                "rustls_server_config_free: invariant failed: arc.strong_count was < 1: {}. \
+                You must not free the same server_config multiple times.",
+                strong_count
+            );
```

Comment:
I forget if we talked about this on a previous PR, but I vote for `panic` here. If we can detect a double free, the best thing to do is to crash to make it as easy as possible for the programmer who misused `rustls_server_config_free` to debug the problem.

Reply:
`panic`'ing across the C-to-rust boundary is undefined behavior, so I want to avoid it. Some discussion at https://users.rust-lang.org/t/feedback-on-c-to-rustls-bindings/54201/6. A related alternative would be to abort instead, but Daniel Stenberg has correctly pointed out that libraries should never abort, but should leave the choice to their caller.

That said, I've been convinced that this branch is not actually meaningful. If strong_count is 0, `Arc`'s internal invariants mean it will already have dropped. Once it's dropped, accessing the related memory (including the strong_count) is, I think, undefined. So I think this is just one of those cases where we are unavoidably dependent on the C caller not to to invoke undefined behavior by freeing something multiple times. I'd be interested in ideas on how we could be robust to multiple frees, though!

Reply:
That's what I'd forgotten, that `curl` has rules about libraries crashing intentionally. I have no objection to complying with that requirement, and your point about UB makes it moot anyway.